### PR TITLE
Handle GitHub's ReviewDismissed events and events in dismissed state

### LINKED ISSUE #8370: Handle status errors gracefully on campaign views
When a status that is not known is encountered, we should still show the information we have and only show an error in the component/attribute that fails. A single error in a change should not block or error the entire campaign view completely.

A customer (https://app.hubspot.com/contacts/2762526/company/557475882) encountered this error when a GitHub PR was moved into the dismissed state.

image

### PR DESCRIPTION
**Pull request by mrnugget**
This fixes #8370 by handling GitHub's support for dismissing reviews.

There are two ways in which the dismissal of a review can show up in our codebase.

1. As a `ReviewDismissed` event. This is emitted when a review is dimissed (either by the author of the reviewD itself or by another person)
2. As the `DISMISSED` state of a review. When a review is dimissed, the `PullRequestReview` object embedded in a `PullRequestReviewedEvent` is updated and its state is changed to `DISMISSED`.

Here's what these two events look like in our database:

```
-[ RECORD 5 ]+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
id           | 23022
changeset_id | 12
kind         | github:reviewed
key          | 357299055
created_at   | 2020-02-12 01:51:10.305031-08
metadata     | {"URL": "https://github.com/sourcegraph/automation-testing/pull/156#pullrequestreview-357299055", "Body": "looks good", "State": "DISMISSED", "Author": {"URL": "https://github.com/mrnugget", "Login": "mrnugget", "AvatarURL": "https://avatars3.githubusercontent.com/u/1185253?v=4"}, "Commit": {"OID": "05e562cc5d8ab50b5fcab8dfd5dfa5db13a48e80", "URL": "https://github.com/sourcegraph/automation-testing/commit/05e562cc5d8ab50b5fcab8dfd5dfa5db13a48e80", "Status": {"State": "", "Contexts": null}, "Message": "Add foobar.md file", "Committer": {"Name": "Thorsten Ball", "User": {"URL": "https://github.com/mrnugget", "Login": "mrnugget", "AvatarURL": "https://avatars3.githubusercontent.com/u/1185253?v=4"}, "Email": "mrnugget@gmail.com", "AvatarURL": "https://avatars3.githubusercontent.com/u/1185253?v=4"}, "PushedDate": "2019-11-12T13:00:07Z", "CommittedDate": "2019-11-12T12:57:33Z", "MessageHeadline": "Add foobar.md file"}, "CreatedAt": "2020-02-12T09:28:18Z", "UpdatedAt": "2020-02-12T09:28:33Z", "DatabaseID": 357299055, "AuthorAssociation": "MEMBER", "IncludesCreatedEdit": false}
updated_at   | 2020-02-12 04:22:57.021974-08
-[ RECORD 6 ]+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
id           | 23023
changeset_id | 12
kind         | github:review_dismissed
key          | ryanslade:357299055:1581499713000000000
created_at   | 2020-02-12 01:51:10.305031-08
metadata     | {"Actor": {"URL": "https://github.com/ryanslade", "Login": "ryanslade", "AvatarURL": "https://avatars3.githubusercontent.com/u/25610?v=4"}, "Review": {"URL": "https://github.com/sourcegraph/automation-testing/pull/156#pullrequestreview-357299055", "Body": "looks good", "State": "DISMISSED", "Author": {"URL": "https://github.com/mrnugget", "Login": "mrnugget", "AvatarURL": "https://avatars3.githubusercontent.com/u/1185253?v=4"}, "Commit": {"OID": "05e562cc5d8ab50b5fcab8dfd5dfa5db13a48e80", "URL": "https://github.com/sourcegraph/automation-testing/commit/05e562cc5d8ab50b5fcab8dfd5dfa5db13a48e80", "Status": {"State": "", "Contexts": null}, "Message": "Add foobar.md file", "Committer": {"Name": "Thorsten Ball", "User": {"URL": "https://github.com/mrnugget", "Login": "mrnugget", "AvatarURL": "https://avatars3.githubusercontent.com/u/1185253?v=4"}, "Email": "mrnugget@gmail.com", "AvatarURL": "https://avatars3.githubusercontent.com/u/1185253?v=4"}, "PushedDate": "2019-11-12T13:00:07Z", "CommittedDate": "2019-11-12T12:57:33Z", "MessageHeadline": "Add foobar.md file"}, "CreatedAt": "2020-02-12T09:28:18Z", "UpdatedAt": "2020-02-12T09:28:33Z", "DatabaseID": 357299055, "AuthorAssociation": "MEMBER", "IncludesCreatedEdit": false}, "CreatedAt": "2020-02-12T09:28:33Z", "DismissalMessage": "Things"}
updated_at   | 2020-02-12 04:22:57.021974-08
```

The code here handles *both* cases.

### CODE REVIEW

**Comment by mrnugget** [comment]
The diff might look noisy but that's mostly because I've removed a lot of the redundancy in how the author or the state of a review-related event was calculated. Now we only have one place: `ReviewState` and `ReviewAuthor` methods on `ChangesetEvent`. That's used by both `ChangesetEvents` (plural!) `ReviewState` method and by the `CalcCounts` function.

File: enterprise/internal/a8n/counts.go [hunk 1]
```diff
@@ -237,17 +246,32 @@ func computeCounts(c *ChangesetCounts, csEvents Events) error {
...
+			if author == "" {
+				continue
+			}
+
+			if e.Type() == a8n.ChangesetEventKindBitbucketServerUnapproved {
+				// A BitbucketServer Unapproved can only follow a previous Approved by
+				// the same author.
+				lastReview, ok := lastReviewByAuthor[author]
+				if !ok || lastReview != a8n.ChangesetReviewStateApproved {
+					return errors.New("Bitbucket Server Unapproval not following an Approval")
```
Comment (ryanslade) [inline enterprise/internal/a8n/counts.go]: Do we want this to be a hard error as I think it'll bubble up to the frontend

File: enterprise/internal/a8n/counts.go [hunk 2]
```diff
@@ -237,17 +246,32 @@ func computeCounts(c *ChangesetCounts, csEvents Events) error {
...
-			// the same author.
-			lastReview, ok := lastReviewByAuthor[author]
-			if !ok || lastReview != a8n.ChangesetReviewStateApproved {
-				return errors.New("Bitbucket Server Unapproval not following an Approval")
+			if e.Type() == a8n.ChangesetEventKindGitHubReviewDismissed {
+				// A GitHub Review Dismissed can only follow a previous review by
+				// the author of the review included in the event.
+				_, ok := lastReviewByAuthor[author]
+				if !ok {
+					return errors.New("GitHub review dismissal not following a review")
```
Comment (ryanslade) [inline enterprise/internal/a8n/counts.go]: Same as previous comment

File: enterprise/internal/a8n/counts.go (hunk 1 above)
Comment (mrnugget) [reply enterprise/internal/a8n/counts.go]: That's been here before, but okay. Let's turn this into a warning log and `continue`.

**Pull request merged**